### PR TITLE
Add data fetching and processing functions in app.py

### PR DESCRIPTION
**Pull request by toqaalaa20**

### CODE REVIEW

**Review by toqaWad — CHANGES_REQUESTED** [review]
Review of Pull Request #1: "Add data fetching and processing functions in app.py"

Overall, the PR introduces new functionality as described. However, there are several areas for improvement based on our code review checklist.

### Warnings and Recommendations:

**1. Security Considerations:**
*   **app.py, Lines 4-5 (`getData` function):**
    *   **Warning:** Potential lack of authentication/authorization and secure data handling for external API calls, especially if the API returns sensitive data.
    *   **Recommendation:** Implement proper authentication mechanisms (e.g., API keys, OAuth tokens) and ensure sensitive data is handled securely (e.g., encryption at rest/in transit).
*   **app.py, Lines 7-8 (`calc` function):**
    *   **Warning:** Lack of input validation for `x` and `y`.
    *   **Recommendation:** Validate inputs to `calc` to ensure they are of the expected type and within reasonable bounds to prevent unexpected behavior.
*   **app.py, Lines 13-14 (`process` function):**
    *   **Warning:** Lack of input validation/error handling for the structure of `data` received from `getData`.
    *   **Recommendation:** Add checks to ensure `d["active"]` and `d["value"]` keys exist before access to prevent `KeyError`.

**2. Code Readability & Maintainability:**
*   **app.py, Lines 3, 7, 12 (Functions `getData`, `calc`, `process`):**
    *   **Warning:** Missing docstrings for functions.
    *   **Recommendation:** Add clear docstrings explaining the purpose, arguments, and return values of each function.
*   **app.py, Line 4 (`getData` function):**
    *   **Warning:** Variable name `r` is not very descriptive.
    *   **Recommendation:** Consider renaming `r` to `response` for clarity.
*   **app.py, Line 13 (`process` function):**
    *   **Warning:** Variable name `d` is not very descriptive.
    *   **Recommendation:** Consider renaming `d` to `item` or `record` for clarity.

**3. Testing & Validation:**
*   **app.py (General):**
    *   **Warning:** Insufficient test coverage.
    *   **Recommendation:** Add unit tests for `getData`, `calc`, and `process` functions to ensure correctness and prevent regressions, covering both intended functionality and edge cases (e.g., API failures, malformed data).

**4. Error Handling & Resilience:**
*   **app.py, Lines 4-5 (`getData` function):**
    *   **Warning:** Lack of robust error handling for network requests (e.g., connection errors, HTTP status codes other than 2xx) and JSON parsing errors.
    *   **Recommendation:** Implement `try-except` blocks to gracefully handle `requests.exceptions.RequestException` and `json.JSONDecodeError`. Consider retries or a circuit breaker pattern for external API calls.
*   **app.py, Lines 13-14 (`process` function):**
    *   **Warning:** Lack of error handling for missing keys (e.g., "active", "value") when processing `data`.
    *   **Recommendation:** Use `d.get("key")` with a default value or explicit `try-except KeyError` blocks when accessing dictionary keys that might not always be present.

Please address these points to improve the quality, robustness, and maintainability of the code.

File: app.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+def calc(x, y):
+    return x + y * 2
+
+config = {"mode": "prod", "retries": 3}
+
+def process():
+    data = getData()
+    for d in data:
+        if d["active"]:
```
Comment (toqaWad) [inline app.py:15]: Warning: The `process` function directly uses `d["value"]` from the external API response in calculations without explicit input validation. Consider adding validation to prevent potential security vulnerabilities.

File: app.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import requests
+
+def getData():
+    r = requests.get("https://api.example.com/data")
```
Comment (toqaWad) [inline app.py:4]: Warning: The `getData` function makes a request to an external API without any visible authentication mechanism. Please ensure sensitive data and APIs are properly secured.

File: app.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    return r.json()
```
Comment (toqaWad) [inline app.py:5]: Warning: The `getData` function lacks robust error handling for the network request. If `requests.get()` fails or returns a non-200 status, `r.json()` could raise an exception. Implement `try-except` blocks for network operations.

File: app.py
Comment (toqaWad) [inline app.py:1]: Warning: There are no unit or integration tests provided for the new functions in `app.py`. Please add sufficient test coverage to validate functionality and handle edge cases.

**Review by toqaWad — CHANGES_REQUESTED** [review]
The following warnings were identified based on the code review checklist:

- **Security Considerations**: The `getData` function makes an external API request without a clear authentication mechanism, and `d["value"]` is used in calculations without explicit input validation, posing potential security risks.

- **Testing & Validation**: There is a lack of unit or integration tests for the new functions, which is crucial for verifying functionality and handling edge cases.

- **Error Handling & Resilience**: The `getData` function does not include robust error handling for network requests, which could lead to exceptions if the API call fails.

Please address these concerns to improve the security, reliability, and maintainability of the code.
